README.md
English | [简体中文](./README-CN.md)


<p align="center">
<a href=" https://www.alibabacloud.com"><img src="https://aliyunsdk-pages.alicdn.com/icons/AlibabaCloud.svg"></a>
</p>

<h1 align="center">Alibaba Cloud KMS SDK for PHP</h1>

<p align="center">
<a href="https://packagist.org/packages/alibabacloud/kms"><img src="https://poser.pugx.org/alibabacloud/kms/v/stable" alt="Latest Stable Version"></a>
<a href="https://packagist.org/packages/alibabacloud/kms"><img src="https://poser.pugx.org/alibabacloud/kms/v/unstable" alt="Latest Unstable Version"></a>
<a href="https://packagist.org/packages/alibabacloud/kms"><img src="https://poser.pugx.org/alibabacloud/kms/composerlock" alt="composer.lock"></a>
<a href="https://packagist.org/packages/alibabacloud/kms"><img src="https://poser.pugx.org/alibabacloud/kms/downloads" alt="Total Downloads"></a>
<a href="https://packagist.org/packages/alibabacloud/kms"><img src="https://poser.pugx.org/alibabacloud/kms/license" alt="License"></a>
<br/>
<a href="https://codecov.io/gh/aliyun/php-kms-sdk"><img src="https://codecov.io/gh/aliyun/php-kms-sdk/branch/master/graph/badge.svg" alt="codecov"></a>
<a href="https://scrutinizer-ci.com/g/aliyun/php-kms-sdk"><img src="https://scrutinizer-ci.com/g/aliyun/php-kms-sdk/badges/quality-score.png" alt="Scrutinizer Code Quality"></a>
<a href="https://travis-ci.org/aliyun/php-kms-sdk"><img src="https://travis-ci.org/aliyun/php-kms-sdk.svg?branch=master" alt="Travis Build Status"></a>
<a href="https://ci.appveyor.com/project/aliyun/php-kms-sdk/branch/master"><img src="https://ci.appveyor.com/api/projects/status/9n1qmw9i5qxx507t/branch/master?svg=true" alt="Appveyor Build Status"></a>
<a href="https://scrutinizer-ci.com/code-intelligence"><img src="https://scrutinizer-ci.com/g/aliyun/php-kms-sdk/badges/code-intelligence.svg" alt="Code Intelligence Status"></a>
</p>


Alibaba Cloud KMS SDK for PHP Supports PHP developers using Alibaba Cloud [Key Management Service (KMS)](https://kms.console.aliyun.com/).


## Prerequisites
Your system will need to meet the [Prerequisites](docs/0-Prerequisites-EN.md), including having PHP >= 5.5. We highly recommend having it compiled with the cURL extension and cURL 7.16.2+.


## Installation
If Composer is already [installed globally on your system](https://getcomposer.org/doc/00-intro.md#globally), run the following in the base directory of your project to install Alibaba Cloud KMS SDK for PHP as a dependency:
```
composer require alibabacloud/kms
```
> Some users may not be able to install due to network problems, you can try to switch the Composer mirror.

Please see the [Installation](docs/1-Installation-EN.md) for more detailed information about installing through Composer and other means.


## Setting up the client
Before you begin, you need to sign up for an Alibaba Cloud account and retrieve your [Credentials](https://usercenter.console.aliyun.com/#/manage/ak).
Please pass in your `accessKeyId`, `accessKeySecret`, `endpoint`, [View the list of KMS endpoints](https://developer.aliyun.com/endpoints#service_kms).

```php
<?php

use AlibabaCloud\Kms\V20160120\Kms;

$client = new Kms('accessKeyId', 'accessKeySecret', 'kms.cn-hangzhou.aliyuncs.com');
```


## Set request options
Please refer to [Guzzle Request Options][guzzle-docs].

```php
<?php

$options = [];

```


## Cancel Key Deletion
Cancels the deletion of a CMK. When this operation is successful, the CMK is set to the Enabled state. [API Reference](https://www.alibabacloud.com/help/doc-detail/44197.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->cancelKeyDeletion(
        [
            'KeyId' => 'key_id',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Create Alias
Creates a display name for a CMK. [API Reference](https://www.alibabacloud.com/help/doc-detail/68624.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->createAlias(
        [
            'AliasName' => 'alias/1234',
            'KeyId'     => 'key_id',
        ]
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Create Key
Creates a customer master key (CMK). [API Reference](https://www.alibabacloud.com/help/doc-detail/28947.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->createKey(
        [
            'Origin'      => 'Aliyun_KMS',
            'Description' => 'test key',
            'KeyUsage'    => 'ENCRYPT/DECRYPT',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}
```


## Decrypt
Decrypts ciphertext. Ciphertext is plaintext that has been previously encrypted by using any of the following operations. [API Reference](https://www.alibabacloud.com/help/doc-detail/28950.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->decrypt(
        [
            'CiphertextBlob'    => 'CiphertextBlob',
            'EncryptionContext' => json_encode(['k' => 'v']),
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}
```


## Delete Alias
Deletes the specified alias. [API Reference](https://www.alibabacloud.com/help/doc-detail/68626.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->deleteAlias(
        [
            'AliasName' => 'alias/12345',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}


```

## Delete Key Material
Deletes the imported key material. [API Reference](https://www.alibabacloud.com/help/doc-detail/68623.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->deleteKeyMaterial(
        [
            'KeyId' => 'id',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Describe Key
Returns detailed information about the specified CMK. [API Reference](https://www.alibabacloud.com/help/doc-detail/28952.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->describeKey(
        [
            'KeyId' => 'id',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Describe Regions
Returns available regions for the specified account. [API Reference](https://www.alibabacloud.com/help/doc-detail/54560.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->describeRegions();
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Disable Key
Sets the state of a CMK to disabled, thereby preventing its use for cryptographic operations. The ciphertext encrypted using the CMK cannot be decrypted until the CMK is enabled again. [API Reference](https://www.alibabacloud.com/help/doc-detail/35151.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->disableKey(
        [
            'KeyId' => 'id',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Enable Key
Sets the state of the specified CMK to Enabled, thereby permitting its use for cryptographic operations. [API Reference](https://www.alibabacloud.com/help/doc-detail/35150.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->enableKey(
        [
            'KeyId' => 'id',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Encrypt
Encrypts plaintext into ciphertext by using a CMK. [API Reference](https://www.alibabacloud.com/help/doc-detail/28949.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->encrypt(
        [
            'KeyId'             => 'id',
            'Plaintext'         => 'text',
            'EncryptionContext' => json_encode(['k' => 'v']),
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Generate Data Key
Returns a data encryption key that you can use in your application to encrypt data locally. [API Reference](https://www.alibabacloud.com/help/doc-detail/28948.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->generateDataKey(
        [
            'KeyId' => 'id',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Get Parameters for Import
Returns the items you need in order to import key material into KMS from your existing key management infrastructure. The returned items are used in the subsequent ImportKeyMaterial request. [API Reference](https://www.alibabacloud.com/help/doc-detail/68621.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->getParametersForImport(
        [
            'KeyId'             => 'external_key_id',
            'WrappingAlgorithm' => 'RSAES_OAEP_SHA_256',
            'WrappingKeySpec'   => 'RSA_2048',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Import Key Material
Imports key material to an existing KMS CMK that was created without key material. [API Reference](https://www.alibabacloud.com/help/doc-detail/68622.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->importKeyMaterial(
        [
            'KeyId'                 => 'external_key_id',
            'EncryptedKeyMaterial'  => base64_encode('test'),
            'ImportToken'           => 'import_token',
            'KeyMaterialExpireUnix' => time() + 3600,
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## List Aliases
Gets a list of all aliases in the caller’s account and region. [API Reference](https://www.alibabacloud.com/help/doc-detail/68627.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->listAliases(
        [
            'PageNumber' => 1,
            'PageSize'   => 100,
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## List Aliases by KeyId
Lists all aliases associated with the CMK. [API Reference](https://www.alibabacloud.com/help/doc-detail/68628.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->listAliasesByKeyId(
        [
            'KeyId'      => 'key_id',
            'PageNumber' => 1,
            'PageSize'   => 100,
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## List Keys
Returns a list of all CMKs in the caller’s account and region. [API Reference](https://www.alibabacloud.com/help/doc-detail/28951.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->listKeys(
        [
            'PageNumber' => 1,
            'PageSize'   => 100,
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Schedule Key Deletion
Schedules the deletion of a CMK. [API Reference](https://www.alibabacloud.com/help/doc-detail/44196.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->scheduleKeyDeletion(
        [
            'KeyId'               => 'key_id',
            'PendingWindowInDays' => 7,
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Update Alias
Associates an existing alias with a different CMK. [API Reference](https://www.alibabacloud.com/help/doc-detail/68625.htm)

```php
<?php

use GuzzleHttp\Exception\GuzzleException;

try {
    $result = $client->updateAlias(
        [
            'KeyId'     => 'key_id',
            'AliasName' => 'alias/12345',
        ],
        $options
    );
} catch (GuzzleException $e) {
    echo $e->getMessage();
}

```


## Issues
[Opening an Issue](https://github.com/aliyun/php-kms-sdk/issues/new/choose), Issues not conforming to the guidelines may be closed immediately.


## Changelog
Detailed changes for each release are documented in the [release notes](CHANGELOG.md).


## Contribution
Please make sure to read the [Contributing Guide](CONTRIBUTING.md) before making a pull request.


## References
* [OpenAPI Explorer][open-api]
* [Packagist][packagist]
* [Composer][composer]
* [Guzzle Documentation][guzzle-docs]
* [Latest Release][latest-release]


## License
[Apache-2.0](LICENSE.md)

Copyright 1999-2019 Alibaba Group Holding Ltd.


[open-api]: https://api.aliyun.com
[latest-release]: https://github.com/aliyun/php-kms-sdk
[guzzle-docs]: https://guzzle-cn.readthedocs.io/zh_CN/latest/request-options.html
[composer]: http://getcomposer.org
[packagist]: https://packagist.org/packages/alibabacloud/kms
[client]: https://github.com/aliyun/openapi-sdk-php-client/blob/master/README-CN.md#alibaba-cloud-client-for-php
[clients]: https://github.com/aliyun/openapi-sdk-php-client/blob/master/docs/2-Client-EN.md
[request]: https://github.com/aliyun/openapi-sdk-php-client/blob/master/docs/3-Request-EN.md
[result]: https://github.com/aliyun/openapi-sdk-php-client/blob/master/docs/4-Result-EN.md
[ak]: https://usercenter.console.aliyun.com/?spm=5176.doc52740.2.3.QKZk8w#/manage/ak
[home]: https://home.console.aliyun.com/?spm=5176.doc52740.2.4.QKZk8w
[cURL]: http://php.net/manual/en/book.curl.php
[OPCache]: http://php.net/manual/en/book.opcache.php
[xdebug]: http://xdebug.org
[OpenSSL]: http://php.net/manual/en/book.openssl.php
[aliyun]: https://www.aliyun.com
[alibabacloud]: https://www.alibabacloud.com


CHANGELOG.md
# CHANGELOG

## 1.1.0 - 2019-04-01
- Updated `composer.json`.


## 1.0.5 - 2019-03-26
- Updated Description.


## 1.0.4 - 2019-03-25
- Updated README.md.


## 1.0.3 - 2019-03-23
- Remove Svg.


## 1.0.2 - 2019-03-19
- Update Docs.


## 1.0.1 - 2019-03-13
- Update Docs.


## 1.0.0 - 2019-01-22

- Initial release of the Alibaba Cloud KMS SDK for PHP Version 1.0.0 on Packagist See <https://github.com/aliyun/php-kms-sdk> for more information.


tests/Feature/KmsWithOptionsTest.php
<?php

namespace AlibabaCloud\Kms\Tests\Feature;

use AlibabaCloud\Kms\V20160120\Kms;
use GuzzleHttp\Exception\GuzzleException;
use PHPUnit\Framework\TestCase;

/**
 * Class KmsWithOptionsTest
 *
 * @package AlibabaCloud\Kms\Tests\Feature
 */
class KmsWithOptionsTest extends TestCase
{
    /**
     * @var Kms
     */
    protected $client;

    /**
     * @var array
     */
    protected $runtime = [
        'connect_timeout' => 0.1,
        'timeout'         => 0.1,
    ];

    public function setUp()
    {
        parent::setUp();
        $this->client = new Kms(
            getenv('ACCESS_KEY_ID'),
            getenv('ACCESS_KEY_SECRET'),
            getenv('ENDPOINT')
        );
    }

    public function testDescribeRegions()
    {
        try {
            $this->client->describeRegions([], $this->runtime);
        } catch (GuzzleException $e) {
            self::assertStringStartsWith('cURL error', $e->getMessage());
        }
    }
}


src/V20160120/Kms.php
<?php

namespace AlibabaCloud\Kms\V20160120;

use AlibabaCloud\Client\Result\Result;
use AlibabaCloud\Kms\Base;
use GuzzleHttp\Exception\GuzzleException;

/**
 * Class Kms
 *
 * @package AlibabaCloud\V20160120\\Kms
 */
class Kms extends Base
{
    /**
     * @var string
     */
    protected $version = '2016-01-20';

    /**
     * @description cancel key deletion which can enable this key
     * @see         https://help.aliyun.com/document_detail/44197.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function cancelKeyDeletion($query = [], $runtime = [])
    {
        $query['Action']     = 'CancelKeyDeletion';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description make alias to key
     * @see         https://help.aliyun.com/document_detail/68624.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - AliasName string required: cmk alias, prefix must be 'alias/'
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function createAlias($query = [], $runtime = [])
    {
        $query['Action']     = 'CreateAlias';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description create new key
     * @see         https://help.aliyun.com/document_detail/28947.html
     *
     * @param $query
     *   - Action string required
     *   - Origin string optional: Aliyun_KMS (default) or EXTERNAL
     *   - Description string optional: description of key
     *   - KeyUsage string optional: usage of key, default is ENCRYPT/DECRYPT
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function createKey($query = [], $runtime = [])
    {
        $query['Action']     = 'CreateKey';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description decrypt body of CiphertextBlob
     * @see         https://help.aliyun.com/document_detail/28950.html
     *
     * @param $query
     *   - Action string required
     *   - CiphertextBlob string required: ciphertext to be decrypted.
     *   - EncryptionContext string optional: key/value string, must be {string: string}
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function decrypt($query = [], $runtime = [])
    {
        $query['Action']     = 'Decrypt';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description delete alias
     * @see         https://help.aliyun.com/document_detail/68626.html
     *
     * @param $query
     *   - Action string required
     *   - AliasName string required: alias name, prefix must be 'alias/'
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function deleteAlias($query = [], $runtime = [])
    {
        $query['Action']     = 'DeleteAlias';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description delete key material
     * @see         https://help.aliyun.com/document_detail/68623.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function deleteKeyMaterial($query = [], $runtime = [])
    {
        $query['Action']     = 'DeleteKeyMaterial';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description get description of main key
     * @see         https://help.aliyun.com/document_detail/28952.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function describeKey($query = [], $runtime = [])
    {
        $query['Action']     = 'DescribeKey';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description query available regions
     * @see         https://help.aliyun.com/document_detail/54560.html
     *
     * @param $query
     *   - Action string required
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function describeRegions($query = [], $runtime = [])
    {
        $query['Action']     = 'DescribeRegions';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description disable key
     * @see         https://help.aliyun.com/document_detail/35151.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function disableKey($query = [], $runtime = [])
    {
        $query['Action']     = 'DisableKey';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description enable key
     * @see         https://help.aliyun.com/document_detail/35150.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function enableKey($query = [], $runtime = [])
    {
        $query['Action']     = 'EnableKey';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description encrypt content
     * @see         https://help.aliyun.com/document_detail/28949.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - Plaintext string required: plaintext to be encrypted (must be Base64 encoded)
     *   - EncryptionContext string optional: key/value string, must be {string: string}
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function encrypt($query = [], $runtime = [])
    {
        $query['Action']     = 'Encrypt';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description generate local data key
     * @see         https://help.aliyun.com/document_detail/28948.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - KeySpec string optional: AES_256 or AES_128
     *   - NumberOfBytes int optional: length of key
     *   - EncryptionContext string optional: key/value string, must be {string: string}
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function generateDataKey($query = [], $runtime = [])
    {
        $query['Action']     = 'GenerateDataKey';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description get the imported master key (CMK) material
     * @see         https://help.aliyun.com/document_detail/68621.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - WrappingAlgorithm string required: algorithm for encrypting key material, RSAES_PKCS1_V1_5, RSAES_OAEP_SHA_1
     *   or RSAES_OAEP_SHA_256
     *   - WrappingKeySpec string required: public key type used to encrypt key material, RSA_2048
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function getParametersForImport($query = [], $runtime = [])
    {
        $query['Action']     = 'GetParametersForImport';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description
     * @see https://help.aliyun.com/document_detail/68622.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - EncryptedKeyMaterial string required: key material encrypted with base64
     *   - ImportToken string required: obtained by calling GetParametersForImport
     *   - KeyMaterialExpireUnix {timestamp} optional: Key material expiration time
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function importKeyMaterial($query = [], $runtime = [])
    {
        $query['Action']     = 'ImportKeyMaterial';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description returns all aliases of the current user in the current zone
     * @see         https://help.aliyun.com/document_detail/68627.html
     *
     * @param $query
     *   - Action string required
     *   - PageNumber int optional: current page, default 1
     *   - PageSize int optional: result count (0 - 100), default 10
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function listAliases($query = [], $runtime = [])
    {
        $query['Action']     = 'ListAliases';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description list all aliases corresponding to the specified master key (CMK)
     * @see         https://help.aliyun.com/document_detail/68628.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - PageNumber int optional: current page, default 1
     *   - PageSize int optional: result count (0 - 100), default 10
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function listAliasesByKeyId($query = [], $runtime = [])
    {
        $query['Action']     = 'ListAliasesByKeyId';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description Returns all the master key IDs of the caller in the calling area
     * @see         https://help.aliyun.com/document_detail/28951.html
     *
     * @param $query
     *   - Action string required
     *   - PageNumber int optional: current page, default 1
     *   - PageSize int optional: result count (0 - 100), default 10
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function listKeys($query = [], $runtime = [])
    {
        $query['Action']     = 'ListKeys';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description request to delete a specified master key (CMK)
     * @see         https://help.aliyun.com/document_detail/44196.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - PendingWindowInDays int required: key pre-delete cycle, [7, 30]
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function scheduleKeyDeletion($query = [], $runtime = [])
    {
        $query['Action']     = 'ScheduleKeyDeletion';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }

    /**
     * @description update the master key (CMK) represented by an existing alias
     * @see         https://help.aliyun.com/document_detail/68625.html
     *
     * @param $query
     *   - Action string required
     *   - KeyId string required: global unique identifier
     *   - AliasName string required: the alias to be operated, prefix must be 'alias/'
     * @param $runtime
     *
     * @return Result
     * @throws GuzzleException
     */
    public function updateAlias($query = [], $runtime = [])
    {
        $query['Action']     = 'UpdateAlias';
        $request['protocol'] = 'https';
        $request['method']   = 'GET';
        $request['pathname'] = '/';
        $request['query']    = $this->getQuery($query, $request);
        $request['headers']  = [
            'host' => $this->endpoint,
        ];

        $response = $this->request($request, $runtime);

        return $this->json($response);
    }
}
